pkg/replica/replica.go

```diff
@@ -667,39 +665,38 @@ func (r *Replica) openFile(name string, flag int) (types.DiffDisk, error) {
 	return sparse.NewDirectFileIoProcessor(r.diskPath(name), os.O_RDWR|flag, 06666, true)
 }
 
-func (r *Replica) createNewHead(oldHead, parent, created string, size int64) (f types.DiffDisk, newDisk disk, err error) {
+func (r *Replica) createNewHead(oldHead, parent, created string, size int64) (f types.DiffDisk, newDisk disk, rollbackFunc func() error, err error) {
 	newHeadName, err := r.nextFile(diskPattern, diskutil.VolumeHeadDiskName, oldHead)
 	if err != nil {
-		return nil, disk{}, err
+		return nil, disk{}, nil, err
 	}
 
 	if _, err := os.Stat(r.diskPath(newHeadName)); err == nil {
-		return nil, disk{}, fmt.Errorf("%s already exists", newHeadName)
+		return nil, disk{}, nil, fmt.Errorf("%s already exists", newHeadName)
 	}
 
+	f, err = r.openFile(r.diskPath(newHeadName), os.O_TRUNC)
+	if err != nil {
+		return nil, disk{}, nil, err
+	}
+
+	var subRollbackFunc func() error
 	defer func() {
-		var rollbackErr error
-		if err != nil {
-			if _, err := os.Stat(r.diskPath(newHeadName)); err == nil {
-				if err := os.Remove(r.diskPath(newHeadName)); err != nil {
-					rollbackErr = err
-				}
-			}
-			if _, err := os.Stat(r.diskPath(newHeadName + diskutil.DiskMetadataSuffix)); err == nil {
-				if err := os.Remove(r.diskPath(newHeadName + diskutil.DiskMetadataSuffix)); err != nil {
-					rollbackErr = types.CombineErrors(rollbackErr, err)
-				}
+		rollbackFunc = func() error {
+			f.Close()
```

Comment:
Does `f.Close()` need to be in the `rollbackFunc`?

Reply:
If everything works fine, `f` will be added into `r.volume.files`. In this case, we don't need to close it.
Otherwise, we need to close it before the rollback cleanup.

Reply:
Got it. Thanks for the clarification.